Life cycle of machine learning project :

1. Understand the problem statement <br>
2. Data Collection <br>
3. Data checks to perform <br>
4. EDA <br>
5. Data Pre-Processing <br>
6. Model Training <br>
7. Choose best model <br>

1) Problem Statement:<br>
  * This project understands how the student's performance is affected by other variables such as Gender,Ethnicity,Parental level of education,Lunch and Test preparation course.

2. Data Collection:<br>
  * From Kaggle

2. 1] Import Data and required packages

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [4]:
df=pd.read_csv("data\stud.csv")

In [5]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [6]:
df.shape

(1000, 8)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


2. 2]Dataset Information:

  * Gender-male or female
  * race_ethnicoty-describes which group like group A,B,C
  * parental_level of education-tell about parents degree [high school,associate degree,bachelors,masters]
  * lunch-standard,free/reduced
  * test_preparation-completed or not completed before test
  * Math score
  * Reading score
  * Writing score

3. Data Checks to perform
 * Check Missing values
 * Check Duplicates
 * Check data types
 * Check the number of unique values of each column
 * Check the statistics of data set
 * Check various categories present in the different categorical column

3. 1] Check missing values

In [8]:
df.isnull().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

there is no missing values

3. 2] Check Duplicate value

In [12]:
df.duplicated().sum()

np.int64(0)

3. 3]Check data types

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


3. 4]Check the number of unique values of each column

In [16]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

3. 5]Check the statistics of data set

In [17]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

* From the above description of numerical data,all means are very close to each other-b/w 66 and 68.05 
* All the S.D are close -b/w 14.6 and 15.19
* While there is a minimum score 0 for math,writing minimum is much higher=10 and for reading higher=17

3. 6]Check various categories present in the different categorical column

In [18]:
print("categories in 'gender' column: ",end='')
print(df['gender'].unique())
print("categories in 'race_ethnicity' column: ",end='')
print(df['race_ethnicity'].unique())
print("categories in 'lunch' column: ",end='')
print(df['lunch'].unique())
print("categories in 'parental_level_of_education' column: ",end='')
print(df['parental_level_of_education'].unique())
print("categories in 'test_preparation_course' column: ",end='')
print(df['test_preparation_course'].unique())

categories in 'gender' column: ['female' 'male']
categories in 'race_ethnicity' column: ['group B' 'group C' 'group A' 'group D' 'group E']
categories in 'lunch' column: ['standard' 'free/reduced']
categories in 'parental_level_of_education' column: ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
categories in 'test_preparation_course' column: ['none' 'completed']


In [29]:
df['gender'].dtypes

dtype('O')

In [31]:
#define numerical and categorical features

numerical_features=[feature for feature in df.columns if df[feature].dtype!='O']
categorical_features=[feature for feature in df.columns if df[feature].dtype=='O']

In [32]:
numerical_features

['math_score', 'reading_score', 'writing_score']

In [42]:
print('we have {} numerical features : {}'.format(len(numerical_features),numerical_features))
print('\nwe have {} categorical features : {}'.format(len(categorical_features),categorical_features))

we have 3 numerical features : ['math_score', 'reading_score', 'writing_score']

we have 5 categorical features : ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']
